tests/suites/test_suite_ssl.function

```diff
@@ -5589,4 +5594,88 @@ void cid_sanity( )
 }
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_X509_CRT_PARSE_C:MBEDTLS_USE_PSA_CRYPTO:MBEDTLS_PKCS1_V15:MBEDTLS_SSL_PROTO_TLS1_2:MBEDTLS_RSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_SSL_PROTO_DTLS:MBEDTLS_ENTROPY_C:MBEDTLS_RSA_C:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_CTR_DRBG_C:MBEDTLS_ECP_C */
+void raw_key_agreement_fail( )
+{
+    enum { BUFFSIZE = 17000 };
+    mbedtls_endpoint client, server;
+    mbedtls_psa_stats_t stats;
+    size_t free_slots_before = -1;
+
+#if defined(MBEDTLS_TIMING_C)
+    mbedtls_timing_delay_context timer_client, timer_server;
+#endif
+    uint16_t iana_tls_group_list[] = { MBEDTLS_SSL_IANA_TLS_GROUP_SECP256R1,
+                                       MBEDTLS_SSL_IANA_TLS_GROUP_NONE };
+
+    mbedtls_test_message_queue server_queue, client_queue;
+    mbedtls_test_message_socket_context server_context, client_context;
 
+    mbedtls_message_socket_init( &server_context );
+    mbedtls_message_socket_init( &client_context );
+
+    USE_PSA_INIT( );
+
+    /* Client side, force SECP256R1 to make one key bitflip fail
+     * the raw key agreement. Flipping one bit with a Weierstrass
+     * curve (as opposed to a Montgomery curve) has a high chance of
+     * making it invalid. */
+    TEST_ASSERT( mbedtls_endpoint_init( &client, MBEDTLS_SSL_IS_CLIENT,
+                                        MBEDTLS_PK_RSA, &client_context,
+                                        &client_queue,
+                                        &server_queue, iana_tls_group_list ) == 0 );
+#if defined(MBEDTLS_TIMING_C)
+    mbedtls_ssl_set_timer_cb( &client.ssl, &timer_client,
+                              mbedtls_timing_set_delay,
+                              mbedtls_timing_get_delay );
+#endif
+
+    /* Server side */
+    TEST_ASSERT( mbedtls_endpoint_init( &server, MBEDTLS_SSL_IS_SERVER,
+                                        MBEDTLS_PK_RSA, &server_context,
+                                        &server_queue,
+                                        &client_queue, NULL ) == 0 );
+#if defined(MBEDTLS_TIMING_C)
+    mbedtls_ssl_set_timer_cb( &server.ssl, &timer_server,
+                              mbedtls_timing_set_delay,
+                              mbedtls_timing_get_delay );
+#endif
+
+    TEST_ASSERT( mbedtls_mock_socket_connect( &(client.socket),
+                                              &(server.socket),
+                                              BUFFSIZE ) == 0 );
+
+    TEST_ASSERT( mbedtls_move_handshake_to_state( &(client.ssl),
+                                                  &(server.ssl),
+                                                  MBEDTLS_SSL_CLIENT_KEY_EXCHANGE )
+                 ==  0 );
+
+    mbedtls_psa_get_stats( &stats );
+    /* Save the number of slots in use up to this point.
+     * With PSA, one can be used for the ECDH private key. */
+    free_slots_before = stats.empty_slots;
+
+    /* Force a simulated bitflip in the server key. to make the
```

Comment:
I am looking at this PR as I have been thinking more generally about TLS negative testing. If I understand correctly the goal of this bit flip is to make `psa_raw_key_agreement` returns in error. To do so, we specify a specific curve, move the handshake to a given state, change one well chosen bit in the SSL client context and then resume handshake that will call `psa_raw_key_agreement` that will eventually fail. What about to rather use MBEDTLS_TEST_HOOKS to wrap `psa_raw_key_agreement` thus we can just make it fail when and how we want in the tests ? This has the advantage of being a quite generic approach that could be used with all functions used in TLS implementation. This has some implications on the way the code looks like in the library though and this PR could be the opportunity to see if such changes could be made acceptable. What do you think ?

Reply:
In general, I think adding test hooks that can modify data to some SSL functions would be a good approach to negative testing. (Not an exhaustive approach: it's a good way to reach small modifications of good cases, but not a good way to reach “nonsensical” data.)

However, for this specific case, since the work has already been done on an ad hoc test function, I lean towards keeping it. Let's try out test hooks on the next one.

Reply:
Yeah, test hooks here would be nice, I thought we are limiting their use to SE drivers. I'll keep that in mind for the future :) 
I come from a C++ background, and mocking functions is a common practice for performing tests.
If I had an unlimited workforce, I would propose a general solution for mocking probably _any_ function with minimal burden on library code itself:

1. Add a new build target, let's say "mocks"
2. For this build target, as a prerequisite:

- list all library functions with, for example, [ctags](https://github.com/universal-ctags/ctags);
- generate a file with global data about each function: name, number of calls it had, and potential test info (to be filled in runtime);
- for every function, append "_impl";
- for every function, generate one with the original name in the same translation unit, but doing the standard mock things first - increment call counter, check if current global test data forces any behaviour, and if not, call the original one;

3. Compile;
4. Run unit tests. A unit test is an equivalent of a current test case, but has an associated .json (or .xml) file with information about mock behaviour. An array of structures:

- Function name to mock;
- Call number to mock (specific call number / all);
- Return code;
If needed, this could be extended by:
- Argument expectation (only if argument matches - used to limit all calls only to specific arguments);
- Function name to call with passed arguments - to substitute default behaviour for a different one from mocks, and potentially modify argument internals.

This way we could test any library function mocking any of its internals without adding any code to the library, only unit test code and potentially mocks with different behaviours.
... but that's just an idea :)

Reply:
I am glad I put that on the table as it seems there are ideas about it. There is probably a need to write down what we want to achieve (requirements) with this approach.